src/server/pfs/s3/s3.go

```diff
@@ -109,20 +110,46 @@ func Server(port uint16, driver Driver, clientFactory ClientFactory) (*http.Serv
 	s3Server.Bucket = c
 	s3Server.Object = c
 	s3Server.Multipart = c
-	router := s3Server.Router()
+	return s3Server.Router()
+}
+
+// Server runs an HTTP server with an S3-like API for PFS. This allows you to
+// use s3 clients to access PFS contents.
 
+// Modifications to `routerMap` allow dynamic changes for the mapping between s3 endpoints and API configurations
+func Server(port uint16, defaultRouter *mux.Router, routerMap map[string]*mux.Router, routersLock *sync.Mutex) *http.Server {
+	logger := logrus.WithFields(logrus.Fields{
+		"source": "s3gateway",
+	})
 	server := &http.Server{
 		Addr:         fmt.Sprintf(":%d", port),
 		ReadTimeout:  requestTimeout,
 		WriteTimeout: requestTimeout,
 		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 			// Log that a request was made
 			logger.Infof("http request: %s %s", r.Method, r.RequestURI)
-			router.ServeHTTP(w, r)
+			if strings.HasPrefix(r.Host, "s3-") {
+				routersLock.Lock()
+				defer routersLock.Unlock()
+				host := r.Host
+				if sep := strings.Index(r.Host, "."); sep != -1 {
+					host = host[:sep]
+				}
+				router, ok := routerMap[host]
+				if ok {
+					router.ServeHTTP(w, r)
+				} else {
+					w.WriteHeader(http.StatusInternalServerError)
+				}
+			} else if defaultRouter != nil {
+				defaultRouter.ServeHTTP(w, r)
+			} else {
+				w.WriteHeader(http.StatusInternalServerError)
```

Comment:
These feel more like 400 errors, although I guess there's also the possibility that the worker asks for it before the sidecar has set up the router

Reply:
my thinking here was to the user, who's calling our S3 gateway this is probably more of an internal error of our system not being totally set up yet. 

Specifically, if there's an S3 DNS that hits us but we don't have a backend to support it up yet, that's our bad which makes me think it should be an internal error.

Reply:
Ah okay, that seems reasonable